icefoganalytics/internal-data-portal
# Build Dataset Fields Read Page

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/49

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- https://github.com/icefoganalytics/internal-data-portal/issues/7

# Context

Additionally, make sure the fields tab is not accessible until the user has an approved access request to the dataset, or the dataset is using open access which the user matches.

## Dataset Fields Read Page - /dataset/:slug/fields

### Front-end routes

Various tabs on the same level as this component.

### Back-end routes

* /api/dataset-fields - GET, list data set fields

### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue` - see previous definition
* `DatasetFieldsReadPage.vue`
    * `FieldsTable.vue`
        * Field Name - DatasetField#name
        * Display Name - DatasetField#display_name
        * Field Description - DatasetField#description
        * Data Type - DatasetField#data_type
        * Notes - Dataset#notes

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/f414336c-8c3e-4fb9-8a4a-3e68ff90a377)

# Implementation

Upgrade dynamic tab pattern to permit passing attributes to tab components.
Send "locked" prop to fields tab when user should not be able to read it.
Add `vscode-sql-lit` to list of suggested VS Code plugins, plugin add syntax highlighting to JS format strings tagged as `/* sql */`.
Add policy scoping to dataset-fields controller index endpoint.
Upgrade Sequelize where merge strategy to use "and" instead of overwriting strategy. This is future compatible with Sequelize 7.
Use a standard pattern for query scoping. Pattern is to check for ids included in a tested subquery. This is not the most efficient query pattern, but it is the most repeatable and flexible. If certain queries turn out to be too slow, they will need to be written completely in raw SQL.

# Screenshots

Non-subscribed dataset, restricts "fields" tab access.
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/4af6df8c-3bc2-4c62-a031-ae4e7f5f3783)

Subscribed dataset allows access to "fields" tab
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/e9e308da-0e35-40ad-9b04-af63c3a80bd8)

Dataset fields read-only view (when accessible)
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/74bd364c-684e-4bf4-b388-84426c766872)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. As a Data Owner, create a dataset, with some fields, that has government wide open access.
5. Switch your user type to "User" and check that you can view the new dataset, and see its fields.
6. Switch back to "Data Owner" and create another dataset, with some fields, with _zero_ access grants.
7. Switch your user type to "User" and check that this dataset doesn't show up, and that even if you go to the specific fields tab, the dataset-fields endpoint policy scope prevents you from getting any content back.
8. Switch back to "Data Owner" and create another dataset, with some fields, with an access grant that your low access "User" can subscribe to.
9. Switch your user type to "User" and check that you can see the dataset, and after selecting it, the fields tab is locked.
10. Check that you can subscribe to the dataset, and that after subscription, the fields tab is unlocked and you can see the appropriate fields.
11. Create a new dataset with some fields, and a "screened access" access grant your low access user can request access to.
12. As your low access user, request access to the dataset.
13. Check that the fields tab remains locked.
14. As your dataset owner, approve the access request.
15. Switch back to your low access user, and check that they can now access the "fields" tab.
